assets/src/api.ts

```diff
@@ -66,6 +67,34 @@ export const apiCall = <T>({
       }
     })
 
+export const checkedApiCall = <T, U>({
+  url,
+  dataStruct,
+  parser,
+  defaultResult,
+  fetchArgs,
+}: {
+  url: string
+  dataStruct: Struct<T, any>
+  parser: (data: T) => U
+  defaultResult?: U
+  fetchArgs?: RequestInit
+}): Promise<U> =>
+  fetch(url, fetchArgs)
+    .then(checkResponseStatus)
+    .then((response) => parseJson(response) as { data: unknown })
+    .then(({ data: data }) => {
+      assert(data, dataStruct)
+      return parser(data)
+    })
+    .catch((error) => {
+      if (defaultResult === undefined) {
+        throw error
+      } else {
+        return defaultResult
+      }
```

Comment:
Hmm, agreed that it would ready better, but I share your misgiving about not being sure off the top of my head about whether or not there's some subtle behavior difference.